### PR TITLE
Refactoring to remove uncontroversial language extensions

### PR DESCRIPTION
**Pull request by MaximilianAlgehed**
This PR cleans up a bunch of code by moving (what I consider to be) "uncontroversial" language extensions to the cabal files and removes unnecessary language extensions that have been left over by chance.

### CODE REVIEW

**Comment by MaximilianAlgehed** [comment]
@abailly I'm allergic to long lists of what I consider to be relatively non-controversial language extensions in the file header so I took the liberty to clean that up and move a bunch of stuff to `.cabal` files. Please have a look when you have some spare time to see if there is anything here you find controversial or if there is something I have missed (some language extension that might open up for annoying interactions in the future and therefore shouldn't be in a `.cabal` file).

File: quickcheck-dynamic-iosim/quickcheck-dynamic-iosim.cabal [hunk 1]
```diff
@@ -41,6 +41,20 @@ common lang
...
     StandaloneDeriving
+    DerivingStrategies
+    FlexibleContexts
+    FlexibleInstances
+    GADTs
+    KindSignatures
+    MultiParamTypeClasses
+    TupleSections
+    TypeFamilies
+    TemplateHaskell
```
Comment (abailly-iohk) [inline quickcheck-dynamic-iosim/quickcheck-dynamic-iosim.cabal:52]: Do we really use this?

File: quickcheck-dynamic-iosim/quickcheck-dynamic-iosim.cabal [hunk 2]
```diff
@@ -41,6 +41,20 @@ common lang
...
+    TypeApplications
+    ConstraintKinds
+    DataKinds
+    PolyKinds
+    RankNTypes
 
   ghc-options:
     -Wall -Wnoncanonical-monad-instances -Wunused-packages
```
Comment (abailly-iohk) [inline quickcheck-dynamic-iosim/quickcheck-dynamic-iosim.cabal:60]: While we are at it, how about `-Werror`?

**Review by abailly-iohk — APPROVED** [review]
I share your allergy :) I have left a couple comments, feel free to ignore.

File: quickcheck-dynamic-iosim/quickcheck-dynamic-iosim.cabal (hunk 2 above)
Comment (MaximilianAlgehed) [reply quickcheck-dynamic-iosim/quickcheck-dynamic-iosim.cabal:60]: I vote `-Werror` in CI but not during development?

File: quickcheck-dynamic-iosim/quickcheck-dynamic-iosim.cabal (hunk 1 above)
Comment (MaximilianAlgehed) [reply quickcheck-dynamic-iosim/quickcheck-dynamic-iosim.cabal:52]: That's a good question - I just moved things out of files and removed the most "wtf" ones... I'll have a look at which ones are really necessary and clean the lists up a bit.

**Pull request merged**